<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="generator" content="pandoc">
  <meta name="author" content="Jerome Kelleher, Kevin Thornton, Jaime Ashander, and Peter Ralph (me)">
  <meta name="dcterms.date" content="2018-03-12">
  <title>Efficient pedigree recording for fast population genetics simulation</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="reveal.js/css/reveal.css">
  <style type="text/css">
      code{white-space: pre-wrap;}
      span.smallcaps{font-variant: small-caps;}
      span.underline{text-decoration: underline;}
      div.column{display: inline-block; vertical-align: top; width: 50%;}
  </style>
  <link rel="stylesheet" href="reveal.js/css/theme/simple.css" id="theme">
  <!-- Printing and PDF exports -->
  <script>
    var link = document.createElement( 'link' );
    link.rel = 'stylesheet';
    link.type = 'text/css';
    link.href = window.location.search.match( /print-pdf/gi ) ? 'reveal.js/css/print/pdf.css' : 'reveal.js/css/print/paper.css';
    document.getElementsByTagName( 'head' )[0].appendChild( link );
  </script>
  <!--[if lt IE 9]>
  <script src="reveal.js/lib/js/html5shiv.js"></script>
  <![endif]-->
  <style type="text/css">
  
  .reveal { font-size: 30px; }
  
  .reveal h1 { font-size: 1.5em; } 
  
  .reveal h2 { font-size: 1.2em; } 
  
  .reveal .slides { text-align: left; }
  
  .reveal .slides figure { text-align: center; }
  
  .reveal figcaption { display: none; }
  
  </style>
  \[
  \]
</head>
<body>
  <div class="reveal">
    <div class="slides">

<section id="title-slide">
  <h1 class="title">Efficient pedigree recording for fast population genetics simulation</h1>
  <p class="author">Jerome Kelleher, Kevin Thornton, Jaime Ashander, and Peter Ralph (<em>me</em>)</p>
  <p class="date">12 March 2018</p>
</section>

<section class="slide level2">

<ol type="1">
<li>explain tree sequences and why they are so efficient (5min)</li>
<li>advertise tskit (1min)</li>
<li>recall uses for fwds sims (2min)</li>
<li>explain application to fwds sim recording (5min)</li>
<li>display impressive speedups 2min)</li>
<li>advertise other reasons to have output in tree sequences (2min)</li>
</ol>
<!-- 1. explain tree sequences and why they are so efficient (5min) -->
</section>
<section><section id="the-tree-sequence" class="title-slide slide level1"><h1>The tree sequence</h1></section><section id="history-is-just-a-sequence-of-trees" class="slide level2">
<h2>History is just a sequence of trees</h2>
<figure>
<img data-src="sim_ts.anim.gif" alt="Trees along a chromosome" /><figcaption>Trees along a chromosome</figcaption>
</figure>
</section><section id="arg" class="slide level2">
<h2>ARG?</h2>
<div class="incremental">
<ol type="1">
<li><p>The <em>pedigree</em> (parental relationships) and crossover locations would give us the tree sequence for <em>everyone, ever</em>.</p></li>
<li><p>Much less can fully describe the history relevant to a <em>sample</em> of genomes.</p></li>
<li><p>The information is equivalent to the Ancestral Recombination Graph (ARG).</p></li>
<li><p>ARGs are <em>hard</em>, to infer from data, or to work with analytically.</p></li>
</ol>
</div>
</section><section class="slide level2">

<p><a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004842">Kelleher, Etheridge, and McVean</a> introduced the <strong>tree sequence</strong> data structure to make a fast coalescent simulator.</p>
<ul>
<li><p>stores genealogical <em>and</em> variation data <strong>very</strong> compactly</p></li>
<li><p>efficient algorithms available:</p>
<ul>
<li>calculation of allele frequencies in arbitrary cohorts</li>
<li>linkage disequilibrium</li>
<li>subsetting</li>
<li>log-time haplotype matching</li>
</ul></li>
</ul>
</section><section id="simulated-file-sizes" class="slide level2">
<h2>Simulated file sizes</h2>
<div class="columns">
<div class="column" style="width:50%;">
<figure>
<img data-src="file_size.png" alt="file sizes" /><figcaption>file sizes</figcaption>
</figure>
</div><div class="column" style="width:50%;">
<ul>
<li>HapMap chr1 genetic map (250Mb)</li>
<li>Gutenkunst et al out-of-Africa model (3 pops)</li>
<li>mutation rate <span class="math inline">\(2 \times 10^{-8}\)</span> per gen</li>
<li><p><span class="math inline">\(n=10^7\)</span></p>
<ul>
<li>about 17 million variants</li>
<li>VCF size: 318 TiB (250,000<span class="math inline">\(\times\)</span> larger)</li>
</ul></li>
</ul>
</div>
</div>
</section><section id="example-three-samples-two-trees-two-variant-sites" class="slide level2">
<h2>Example: three samples; two trees; two variant sites</h2>
<figure>
<img data-src="example_tree_sequence.png" alt="Example tree sequence" /><figcaption>Example tree sequence</figcaption>
</figure>
</section><section id="nodes-and-edges" class="slide level2">
<h2>Nodes and edges</h2>
<dl>
<dt>Edges</dt>
<dd><p>Common ancestor (i.e., coalescent) events.</p>
<p>Records: interval (left, right); parent node; child node.</p>
</dd>
<dt>Nodes</dt>
<dd><p>The ancestors those happen in.</p>
<p>Records: time (of birth); ID (implicit).</p>
</dd>
</dl>
</section><section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.0.png" alt="Building a tree sequence" /><figcaption>Building a tree sequence</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.1.png" alt="Building a tree sequence" /><figcaption>Building a tree sequence</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.2.png" alt="Building a tree sequence" /><figcaption>Building a tree sequence</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.3.png" alt="Building a tree sequence" /><figcaption>Building a tree sequence</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.4.png" alt="Building a tree sequence" /><figcaption>Building a tree sequence</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.5.png" alt="Building a tree sequence" /><figcaption>Building a tree sequence</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.6.png" alt="Building a tree sequence" /><figcaption>Building a tree sequence</figcaption>
</figure>
</section><section id="sites-and-mutations" class="slide level2">
<h2>Sites and mutations</h2>
<dl>
<dt>Mutations</dt>
<dd><p>When state changes along the tree.</p>
<p>Records: site it occured at; resulting state; ID (implicit).</p>
</dd>
<dt>Sites</dt>
<dd><p>Where mutations fall on the genome.</p>
<p>Records: genomic position; ancestral (root) state.</p>
</dd>
</dl>
</section><section class="slide level2">

<figure>
<img data-src="sites_muts_walkthrough/sites_muts_walkthrough.0.png" alt="Adding mutations" /><figcaption>Adding mutations</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="sites_muts_walkthrough/sites_muts_walkthrough.1.png" alt="Adding mutations" /><figcaption>Adding mutations</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="sites_muts_walkthrough/sites_muts_walkthrough.2.png" alt="Adding mutations" /><figcaption>Adding mutations</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="sites_muts_walkthrough/sites_muts_walkthrough.3.png" alt="Adding mutations" /><figcaption>Adding mutations</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="sites_muts_walkthrough/sites_muts_walkthrough.4.png" alt="Adding mutations" /><figcaption>Adding mutations</figcaption>
</figure>
</section><section id="succinct-tables" class="slide level2">
<h2>Succinct tables</h2>
<p>A <em>tree sequence</em> is the data structure encoding this sequence of trees (with mutations).</p>
<p>A tree sequence can be stored <em>succinctly</em> using the four tables: <em>nodes</em>, <em>edges</em>, <em>sites</em>, and <em>mutations</em>.</p>
<div class="fragment">
<p>These are stored efficiently (hdf5) on disk with a column for metadata.</p>
<!-- 2. advertise tskit (1min) -->
</div>
</section></section>
<section><section id="tskit" class="title-slide slide level1"><h1>tskit</h1></section><section id="tree-sequence-operations" class="slide level2">
<h2>Tree sequence operations</h2>
<p>We can do these things “very fast”:</p>
<ol type="1">
<li>Read in: tables <span class="math inline">\(\rightarrow\)</span> tree sequence</li>
<li>Write out: tree sequence <span class="math inline">\(\rightarrow\)</span> tables</li>
<li>Iterate over trees,</li>
<li>while computing some statistic (AFS, <span class="math inline">\(\pi\)</span>, <span class="math inline">\(f_4\)</span>, LD, )</li>
<li>Simplify (i.e., subset).</li>
</ol>
<div class="fragment">
<p><em>Upcoming:</em> tools for dealing with tree sequences will be part of <a href="https://github.com/tskit-dev/tskit"><em>tskit</em></a>.</p>
</div>
</section><section id="tsinfer" class="slide level2">
<h2>tsinfer</h2>
<p>What about <em>real</em> data?</p>
<p>Also in progress: <em>tsinfer</em> (Kelleher, McVean) to infer tree sequences from real data.</p>
<p>Video: <a href="https://www.youtube.com/watch?v=MH2b9iU4oUA">Simulating, storing &amp; processing genetic variation data w/millions of samples</a></p>
<!-- 3. recall uses for fwds sims (2min) -->
</section></section>
<section><section id="forwards-simulations-are-useful" class="title-slide slide level1"><h1>Forwards simulations are useful</h1></section><section id="why-forwards-simulations" class="slide level2">
<h2>Why forwards simulations?</h2>
<p>Coalescent simulations are <em>much faster</em> than forwards-time simulations</p>
<div class="fragment">
<p>because they don’t have to keep track of <em>everyone</em>, only the ancestors of your sample.</p>
</div>
<div class="fragment">
<p><strong>But:</strong> selection, or sufficient geographic structure, break the assumptions that coalescent simulations rely on.</p>
</div>
</section><section id="geography-and-selection" class="slide level2">
<h2>Geography, and selection</h2>
<p>So, if you</p>
<ol type="1">
<li>have more than a couple of loci under selection, and/or</li>
<li>have fine enough scale geography that demographic fluctuations are important (e.g., continuous space)</li>
</ol>
<p>then you have to do forwards-time, individual-based simulations.</p>
</section><section id="whole-genomes" class="slide level2">
<h2>Whole genomes?</h2>
<p>To model linked selection, we need chromosome-scale simulations.</p>
<div class="fragment">
<p>Then every individual needs to carry around her genotype (somehow). Even at neutral sites!</p>
</div>
<div class="fragment">
<p>Bummer.</p>
</div>
<div class="fragment">
<p><em>Or not?</em></p>
<!-- 4. explain application to fwds sim recording (5min) -->
</div>
</section></section>
<section><section id="tree-sequence-recording" class="title-slide slide level1"><h1>Tree sequence recording</h1></section><section id="the-main-idea" class="slide level2">
<h2>The main idea</h2>
<p>If we <em>record the tree sequence</em> that relates everyone to everyone else,</p>
<p>after the simulation is over we can put neutral mutations down on the trees.</p>
<div class="fragment">
<p>Since neutral mutations don’t affect demography,</p>
<p>this is <em>equivalent</em> to having kept track of them throughout.</p>
</div>
</section><section class="slide level2">

<p>This will have us recording the entire genetic history of everyone in the population, ever, as we go along.</p>
<div class="fragment">
<p>It is <em>not</em> clear this is a good idea.</p>
</div>
</section><section id="tree-recording-strategy" class="slide level2">
<h2>Tree recording strategy</h2>
<p>To record a tree sequence in real time, every time an individual is born, we must:</p>
<div class="incremental">
<ol type="1">
<li>add each gamete to the Node Table,</li>
<li>add entries to the Edge Table recording which parent each gamete inherited each bit of genome from, and</li>
<li>add any new mutations to the Mutation Table and (if necessary) their locations to the Site Table.</li>
</ol>
</div>
<div class="fragment">
<div class="columns">
<div class="column" style="width:15%;">
<figure>
<img data-src="finger_right.png" alt="Rightarrow" style="width:100.0%" /><figcaption>Rightarrow</figcaption>
</figure>
</div><div class="column" style="width:5%;">

</div><div class="column" style="width:75%;">
<p>This is <em>not</em> a good idea.<br />
It produces <strong>waaaaay</strong> too much data.</p>
</div>
</div>
</div>
</section><section id="simplification" class="slide level2">
<h2>Simplification</h2>
<p><em>Question:</em> given a tree sequence containing the history of many individuals, how do we <em>simplify</em> it to only the history of a subset?</p>
</section><section class="slide level2">

<p>Concretely, given an input tree sequence and a subset of its nodes we call the <em>samples</em>, we want a new tree sequence for which:</p>
<ol type="1">
<li><p>All marginal trees match the corresponding subtree in the input tree sequence.</p></li>
<li><p>Every non-sample vertex in marginal trees have at least two children.</p></li>
<li><p>All nodes and edges ancestral at least one sample.</p></li>
<li><p>No adjacent redundant edges (e.g., <span class="math inline">\((\ell, x, p, c) + (x, r, p, c) \rightarrow (\ell, r, p, c)\)</span>).</p></li>
</ol>
</section><section class="slide level2">

<p><em>Answer:</em> to simplify a tree sequence to the history of the <em>samples</em>:</p>
<ol type="1">
<li><p>Paint each <em>sampled</em> chromosome a distinct color.</p></li>
<li><p>Moving back up the tree sequence, copy colors of each chromosome to the parental chromosomes they inherited from.</p></li>
<li><p>If two colors go in the same spot (<em>coalescence</em>), replace with a new color (unique to that ancestor). Output a node for the ancestor and an edge for the coalescence.)</p></li>
<li><p>Once all colors have coalesced in a given segment, stop propagating it.</p></li>
</ol>
</section><section id="an-example-simplify-these-to-j-and-k" class="slide level2">
<h2>An example: simplify these to J and K</h2>
<figure>
<img data-src="simplify_walkthrough.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.0.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.1.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.2.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.3.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.4.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.5.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.6.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.7.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.8.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.9.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.10.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.11.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.12.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.13.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.14.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.15.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section id="another-example-before-simplification" class="slide level2">
<h2>Another example: before simplification</h2>
<figure>
<img data-src="sim_wf.anim.gif" alt="Wright-Fisher tree sequence" /><figcaption>Wright-Fisher tree sequence</figcaption>
</figure>
</section><section id="another-example-before-simplification-1" class="slide level2">
<h2>Another example: before simplification</h2>
<figure>
<img data-src="sim_wf_unlabeled.anim.gif" alt="Wright-Fisher tree sequence" /><figcaption>Wright-Fisher tree sequence</figcaption>
</figure>
</section><section id="and-after-simplification" class="slide level2">
<h2>… and after simplification</h2>
<figure>
<img data-src="sim_wf_simplified.anim.gif" alt="Simplified Wright-Fisher tree sequence" /><figcaption>Simplified Wright-Fisher tree sequence</figcaption>
</figure>
</section><section id="revised-tree-recording-strategy" class="slide level2">
<h2><em>Revised</em> tree recording strategy</h2>
<p>To record a tree sequence in real time, every time an individual is born, we must:</p>
<ol type="1">
<li>add each gamete to the Node Table,</li>
<li>add entries to the Edge Table recording which parent each gamete inherited each bit of genome from</li>
<li><p>add any new mutations to the Mutation Table and (if necessary) their locations to the Site Table.</p></li>
<li><p>Every so often, <em>simplify</em> the tables stored in memory.</p></li>
</ol>
<!-- 5. display impressive speedups 2min) -->
</section></section>
<section><section id="implementation-and-results" class="title-slide slide level1"><h1>Implementation and results</h1></section><section id="benchmark-implementation" class="slide level2">
<h2>Benchmark implementation</h2>
<ul>
<li><p>Recording, simplifying, and output of tables: <code>C</code> code in <code>msprime</code>. (soon: <code>tskit</code>)</p></li>
<li><p>Simulation: <a href="https://github.com/molpopgen/fwdpp"><code>fwdpp</code></a>, by Kevin Thornton (in <code>C++</code>) (<a href="https://github.com/molpopgen/fwdpy11_arg_example">code</a>)</p></li>
<li><p>Glue: <a href="https://github.com/pybind/pybind11/"><code>pybind11</code></a></p></li>
</ul>
<div class="fragment">
<p><em>Also:</em> a pure <code>python</code> implementation, interfacing with <a href="https://github.com/BoPeng/simuPOP"><code>simuPOP</code></a>.</p>
</div>
</section><section id="simulation-parameters" class="slide level2">
<h2>Simulation parameters</h2>
<ol type="1">
<li>Wright-Fisher population of size <span class="math inline">\(N\)</span></li>
<li>simulated for <span class="math inline">\(10N\)</span> genreations</li>
<li>neutral mutation rate <span class="math inline">\(\mu\)</span> equal to recombination rate <span class="math inline">\(r\)</span> per gamete</li>
<li>many, weakly deleterious mutations: rate <span class="math inline">\(\mu/100\)</span> with <span class="math inline">\(s\)</span> exponentially distributed with mean <span class="math inline">\(2.5/N\)</span>.</li>
</ol>
<div class="fragment">
<p><em>Note:</em> if we recorded tree sequences (“pedigree recording”) then the neutral mutation rate was <em>zero</em> but neutral mutations were added <em>afterwards</em>.</p>
</div>
</section><section class="slide level2">

<p>To compare simulations, the key parameter is the expected <em>total</em> number of crossovers per generation, or: <span class="math display">\[\begin{aligned}
    \rho = 4 N r ,
\end{aligned}\]</span> since this gives the expected number of <em>distinct trees</em> in the tree sequence.</p>
</section><section class="slide level2">

<figure>
<img data-src="rawspeed.png" alt="Total run time per single simulation as a function of region length." style="width:80.0%" /><figcaption>Total run time per single simulation as a function of region length.</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="speedup.png" alt="Relative speedup of simulations" style="width:80.0%" /><figcaption>Relative speedup of simulations</figcaption>
</figure>
</section><section id="memory-use" class="slide level2">
<h2>Memory use</h2>
<p>RAM use is kept down by simplifying as often as you need to.</p>
<!-- 6. advertise other reasons to have output in tree sequences (2min) -->
</section></section>
<section><section id="moving-forward" class="title-slide slide level1"><h1>Moving forward</h1></section><section id="tree-sequences" class="slide level2">
<h2>Tree sequences …</h2>
<ol type="1">
<li><p>are compact, useful ways to store population history <em>including</em> genome sequence.</p></li>
<li><p>can be succinctly encoded in a set of tables, which we provide tools for using.</p></li>
<li><p>can be output during a forwards-time simulation,</p></li>
<li><p>which makes much larger simulations feasible.</p></li>
</ol>
</section><section id="future-uses" class="slide level2">
<h2>Future uses</h2>
<ol type="1">
<li><p>Machine learning needs <em>good</em>, <em>fast</em> simulations to train on.</p></li>
<li><p>Tree sequences allow quick computation of many genomic statistics from real data.</p></li>
</ol>
<div class="fragment">
<p><em>Current work:</em> tree sequence recording in <a href="https://messerlab.org/slim/">SLiM</a> (with Jared Galloway).</p>
</div>
</section></section>
<section><section id="thanks" class="title-slide slide level1"><h1>Thanks</h1></section><section id="acknowledgements" class="slide level2">
<h2>Acknowledgements</h2>
<p>Funding: NSF (PR); Wellcome Trust (JK); NIH (KRT); USF&amp;WS (JDA).</p>
<p>Slides with <a href="http://hakim.se">reveal.js</a> and <a href="https://pandoc.org/">pandoc</a>.</p>
</section></section>
    </div>
  </div>

  <script src="reveal.js/lib/js/head.min.js"></script>
  <script src="reveal.js/js/reveal.js"></script>

  <script>

      // Full list of configuration options available at:
      // https://github.com/hakimel/reveal.js#configuration
      Reveal.initialize({
        // Display the page number of the current slide
        slideNumber: true,
        // Push each slide change to the browser history
        history: true,
        // Transition style
        transition: 'none', // none/fade/slide/convex/concave/zoom
        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // Optional reveal.js plugins
        dependencies: [
          { src: 'reveal.js/lib/js/classList.js', condition: function() { return !document.body.classList; } },
          { src: 'reveal.js/plugin/zoom-js/zoom.js', async: true },
          { src: 'reveal.js/plugin/math/math.js', async: true },
          { src: 'reveal.js/plugin/notes/notes.js', async: true }
        ]
      });
    </script>
    </body>
</html>
